packages/components/src/visually-hidden/component.js

```diff
@@ -10,12 +10,12 @@ import { useVisuallyHidden } from './hook';
  *
  * @example
  * ```jsx
- * import { View, VisuallyHidden } from `@wordpress/components/ui`;
+ * import { VisuallyHidden } from `@wordpress/components`;
  *
  * function Example() {
  * 	return (
  * 		<VisuallyHidden>
- * 			<View as="label">Code is Poetry</View>
+ * 			<label>Code is Poetry</label>
  * 		</VisuallyHidden>
  * 	);
  * }
```

Comment:
I just noticed this `createComponent` util. Looking at its code, it seems it doesn't do much and basically coping that component here would make things simpler for me at least.

What do you think about it? Is it a good abstraction to have?

Reply:
For instance why `useVisuallyHidden` is a hook in the first place, it just adds a class, so why not just add it inline?

Reply:
I think it's a good abstraction to have and encourages "pure hook" components, which I think is generally a good thing (makes us think twice before using `Context` or adding extra components to the React tree for example).

Additionally it's used for a _ton_ of components in `ui` and refactoring them all to just copy the code from `createComponent` would increase the maintenance cost of all those components.

Reply:
> makes us think twice before using Context or adding extra components to the React tree for example

Why is this a bad thing?

Reply:
Additionally, it enforces the use of `View` where we will be able to add back the base styles that we lost by moving away from the custom style system. Because everything is ultimately a View we can just apply the base styles there.

Reply:
> Why is this a bad thing?

We should avoid adding extra components to the React tree as much as possible right? I thought that was a general goal in Gutenberg as a performance consideration? Especially in wp/components where the components are used in a wide range of places, every component added to the tree could add up across the app.

React `Context` providers come with a performance cost as well as adding more components to the tree, right?

Maybe I'm mistaken about this 🤷‍♀️

Reply:
Generally I like the `createComponent` function. Reakit uses a similar abstraction, FWIW, so it's not completely unfounded.

Reply:
> We should avoid adding extra components to the React tree as much as possible right? I thought that was a general goal in Gutenberg as a performance consideration? 

Curious where this guideline comes from :P I don't think this was ever a rule in Gutenberg and I don't really understand how adding components to the tree causes performance regressions.

> React Context providers come with a performance cost as well as adding more components to the tree, right?

Depending how it's used, I can see that yes, and I do agree that Context is generally something not very useful for UI components but I also don't think there's any hard rule there.

Reply:
Now for the createComponent abstraction itself, to be honest, I personally don't like, it's like we're inventing a new way to write React components and feel like it's just a cosmetic abstraction. 

That said, it's internal to the ui components package, so I'm fine keeping it for now. My main worry is that it's an abstraction that doesn't bring much while raising the bar for casual components to understand how components are built in this package, why it's just just a regular component.

Reply:
Well, I think I just picked it up from discussions around the repository and in the WordPress slack 🤔 But maybe I'm imagining things lol! I feel like I remember @gwwar or maybe @ellatrix mentioning something about trying to remove components from the React component tree?

Reply:
Now looking more at createCompnent, It does call `memo` by default in all components, that personally doesn't make sense to me. It's going to do the opposite of what it's meant to solve. Basically, most of these small components don't need to be memoized, React already only update the underlying Dom element if its prop changes, so this extra `memo` is useless. At least we should change the default value to false there.

Reply:
> My main worry is that it's an abstraction that doesn't bring much while raising the bar for casual components to understand how components are built in this package, why it's just just a regular component.

That's fair, it is a new abstraction and does technically redirect from the core functionality of a component.

How do you envision, for example, the `Text` component looking if we removed `createComponent`? Would we get rid of the `useText` hook?

My concern is that if we remove `createComponent`, we'll just be duplicating a ton of code everywhere, indeed, that's why Reakit uses it for example. If a component is follows the pattern established in `createComponent`, namely:

1. Push all props through a hook
2. Pass all processed props to View
3. Connect to the contextSystem

Then we've just duplicated that code, well, more than duplicated, we've replicated probably dozens of times over across the components codebase, just to avoid a tidy, well-tested abstraction. Is that worth it? 🤷‍♀️

Reply:
> How do you envision, for example, the Text component looking if we removed createComponent? Would we get rid of the useText hook?

Well!  for me these `useVisuallyHidden`, `useText`... hooks are very weird hooks. They don't really match what hooks are meant to be used for. They correspond to "components" and not "behavior". It looks like a hack to the hooks system :P 

> Push all props through a hook

What is this exactly? :P Is it a requirement to pass all props through a hook?

> Push all props through a hook

Do all components need to support "as" and other features proposed by the "View" component? I guess my opinion is probably "no" Why should an `Icon` component support that? an Icon always renders an svg (just an example)

> Connect to the contextSystem

Do all components need to connect to the context system. Isn't that just useless overhead for most components? Not sure I understand that system well enough but it seems only useful for components that compose (Card, CardBody...)

Reply:
> At least we should change the default value to false there.

Agreed! We need to update that in `contextConnect` as well.

Reply:
> It looks like a hack to the hooks system :P

That's fair, they probably are 😅

> Is it a requirement to pass all props through a hook?

No, but it's just the way that the g2 components were designed from a technical perspective, to encapsulate all, or as much as possible, logic into a hook that can be reused by other components.

For example, if another component wanted to implement the same logic as `useText` without rendering a `Text` element (like `useHeading` for example, does everything a `Text` can do but with some extras, same with VStack, it's just HStack with some extra props, which in turn is just Flex configured a specific way), this is now possible using these "hacky" hooks.

I'm not saying it's not complicated, but I think approach the technical design of components in this way does add some value around composibility within the library itself (without necessarily exposing all the details of that to consumers of the library).

> Do all components need to support "as" and other features proposed by the "View" component

I think so. It's hard to decide what components _don't_ benefit from the `as` prop. It's an extremely powerful tool for composition.

Icon itself probably doesn't need it, you're correct, but I think the exceptions are few.

> Do all components need to connect to the context system

You'd be surprised. We even have examples of `Text` being accessed via the context system.

I have wondered about the increased overhead of the context system but I think ultimately it's negligible. There were other things like memoization that apparently yielded more significant performance benefits: https://github.com/ItsJonQ/g2/issues/57